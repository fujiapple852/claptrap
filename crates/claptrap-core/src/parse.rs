use crate::clap_ext::IsManyEx;
use crate::output::{CatCmd, ExitCode, Output, Var};
use crate::types::{Command, ValueParser};
use clap::parser::ValuesRef;
use clap::{ArgAction, ColorChoice};
use std::ffi::OsString;
use std::fmt::Display;

/// Parse the provided arguments and generate output.
///
/// This function does not perform any I/O operations.
#[must_use]
#[expect(clippy::needless_pass_by_value)]
pub fn parse<I, T>(cmd: Command, args: I) -> Output
where
    I: IntoIterator<Item = T>,
    T: Into<OsString> + Clone,
{
    let clap_cmd = clap::Command::from(cmd.clone());
    let color = clap_cmd.get_color();
    let help_color = if clap_cmd.is_disable_colored_help_set() {
        ColorChoice::Never
    } else {
        color
    };
    match clap_cmd.clone().try_get_matches_from(args) {
        Ok(matches) => Output::Variables(
            extract_subcommand_path(&matches)
                .into_iter()
                .chain(extract_matches(&cmd, &clap_cmd, &matches, &[]))
                .collect(),
        ),
        Err(err) => match err.kind() {
            clap::error::ErrorKind::DisplayHelp | clap::error::ErrorKind::DisplayVersion => {
                Output::Cat(CatCmd::new(err.render(), ExitCode::Success, help_color))
            }
            clap::error::ErrorKind::DisplayHelpOnMissingArgumentOrSubcommand => {
                Output::Cat(CatCmd::new(err.render(), ExitCode::Usage, help_color))
            }
            _ => Output::Cat(CatCmd::new(err.render(), ExitCode::Error, color)),
        },
    }
}

fn extract_subcommand_path(matches: &clap::ArgMatches) -> Option<Var> {
    let path = std::iter::successors(matches.subcommand(), |&(_, sub)| sub.subcommand())
        .map(|(name, _)| name.to_string())
        .collect::<Vec<_>>();
    (!path.is_empty()).then_some(Var::Subcommand(path))
}

fn extract_matches(
    cmd: &Command,
    clap_cmd: &clap::Command,
    matches: &clap::ArgMatches,
    prefix: &[String],
) -> Vec<Var> {
    let vars = matches.ids().filter_map(|id| {
        let arg = cmd.get_arguments().find(|a| a.get_id() == id.as_str())?;
        let clap_arg = clap_cmd.get_arguments().find(|a| a.get_id() == id)?;
        match clap_arg.get_action() {
            ArgAction::SetTrue | ArgAction::SetFalse => Some(Var::Single(
                prefix.to_vec(),
                id.to_string(),
                matches.get_flag(id.as_str()).to_string(),
            )),
            ArgAction::Count => Some(Var::Single(
                prefix.to_vec(),
                id.to_string(),
                matches.get_count(id.as_str()).to_string(),
            )),
            ArgAction::Append => arg
                .value_parser()
                .and_then(|v| get_many(v, matches, id.as_str()))
                .or_else(|| get_many_raw(matches, id.as_str()))
                .map(|vals| Var::Many(prefix.to_vec(), id.to_string(), vals)),
            ArgAction::Set => {
                if clap_arg.is_many() {
                    arg.value_parser()
                        .and_then(|v| get_many(v, matches, id.as_str()))
                        .or_else(|| get_many_raw(matches, id.as_str()))
                        .map(|vals| Var::Many(prefix.to_vec(), id.to_string(), vals))
                } else {
                    arg.value_parser()
                        .and_then(|v| get_one(v, matches, id.as_str()))
                        .or_else(|| get_one_raw(matches, id.as_str()))
                        .map(|val| Var::Single(prefix.to_vec(), id.to_string(), val))
                }
            }
            _ => None,
        }
    });

    let sub = matches.subcommand().into_iter().flat_map(|(name, sub_m)| {
        cmd.get_subcommands()
            .find(|sc| sc.get_name() == name)
            .into_iter()
            .filter_map(move |sub_cmd| {
                clap_cmd
                    .get_subcommands()
                    .find(|sc| sc.get_name() == name)
                    .map(|clap_sub| {
                        let mut sub_prefix = prefix.to_vec();
                        sub_prefix.push(name.to_string());
                        extract_matches(sub_cmd, clap_sub, sub_m, &sub_prefix)
                    })
            })
            .flatten()
    });

    vars.chain(sub).collect()
}

fn get_one(value_parser: &ValueParser, matches: &clap::ArgMatches, id: &str) -> Option<String> {
    match value_parser {
        ValueParser::Falsey | ValueParser::Boolish | ValueParser::Bool => {
            matches.get_one::<bool>(id).map(ToString::to_string)
        }
        ValueParser::String | ValueParser::PossibleValues(_) => {
            matches.get_one::<String>(id).map(ToString::to_string)
        }
        ValueParser::U8 => matches.get_one::<u8>(id).map(ToString::to_string),
        ValueParser::U16 => matches.get_one::<u16>(id).map(ToString::to_string),
        ValueParser::U32 => matches.get_one::<u32>(id).map(ToString::to_string),
        ValueParser::U64 => matches.get_one::<u64>(id).map(ToString::to_string),
        ValueParser::U128 => matches.get_one::<u128>(id).map(ToString::to_string),
        ValueParser::Usize => matches.get_one::<usize>(id).map(ToString::to_string),
        ValueParser::I8 => matches.get_one::<i8>(id).map(ToString::to_string),
        ValueParser::I16 => matches.get_one::<i16>(id).map(ToString::to_string),
        ValueParser::I32 => matches.get_one::<i32>(id).map(ToString::to_string),
        ValueParser::I64 => matches.get_one::<i64>(id).map(ToString::to_string),
        ValueParser::I128 => matches.get_one::<i128>(id).map(ToString::to_string),
        ValueParser::Isize => matches.get_one::<isize>(id).map(ToString::to_string),
        ValueParser::F32 => matches.get_one::<f32>(id).map(ToString::to_string),
        ValueParser::F64 => matches.get_one::<f64>(id).map(ToString::to_string),
    }
}

fn get_many(
    value_parser: &ValueParser,
    matches: &clap::ArgMatches,
    id: &str,
) -> Option<Vec<String>> {
    fn to_string<'a, T: Display + 'a>(values: ValuesRef<'a, T>) -> Vec<String> {
        values.map(ToString::to_string).collect()
    }
    match value_parser {
        ValueParser::Falsey | ValueParser::Boolish | ValueParser::Bool => {
            matches.get_many::<bool>(id).map(to_string)
        }
        ValueParser::String | ValueParser::PossibleValues(_) => {
            matches.get_many::<String>(id).map(to_string)
        }
        ValueParser::U8 => matches.get_many::<u8>(id).map(to_string),
        ValueParser::U16 => matches.get_many::<u16>(id).map(to_string),
        ValueParser::U32 => matches.get_many::<u32>(id).map(to_string),
        ValueParser::U64 => matches.get_many::<u64>(id).map(to_string),
        ValueParser::U128 => matches.get_many::<u128>(id).map(to_string),
        ValueParser::Usize => matches.get_many::<usize>(id).map(to_string),
        ValueParser::I8 => matches.get_many::<i8>(id).map(to_string),
        ValueParser::I16 => matches.get_many::<i16>(id).map(to_string),
        ValueParser::I32 => matches.get_many::<i32>(id).map(to_string),
        ValueParser::I64 => matches.get_many::<i64>(id).map(to_string),
        ValueParser::I128 => matches.get_many::<i128>(id).map(to_string),
        ValueParser::Isize => matches.get_many::<isize>(id).map(to_string),
        ValueParser::F32 => matches.get_many::<f32>(id).map(to_string),
        ValueParser::F64 => matches.get_many::<f64>(id).map(to_string),
    }
}

fn get_one_raw(matches: &clap::ArgMatches, id: &str) -> Option<String> {
    matches
        .get_raw(id)
        .and_then(|mut values| values.next())
        .map(|value| value.to_string_lossy().into_owned())
}

fn get_many_raw(matches: &clap::ArgMatches, id: &str) -> Option<Vec<String>> {
    matches
        .get_raw(id)
        .map(|values| values.map(|v| v.to_string_lossy().into_owned()).collect())
}
