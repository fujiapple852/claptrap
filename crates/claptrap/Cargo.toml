[package]
name = "claptrap"
description.workspace = true
version.workspace = true
authors.workspace = true
documentation.workspace = true
homepage.workspace = true
repository.workspace = true
keywords.workspace = true
categories.workspace = true
readme = "README.md"
license.workspace = true
edition.workspace = true
rust-version.workspace = true

[dependencies]
claptrap-core.workspace = true
anstream.workspace = true
anyhow.workspace = true
clap = { workspace = true, features = ["derive", "string", "cargo", "wrap_help", "usage", "unstable-styles", "color", "suggestions", "error-context", "env"] }
clap-markdown.workspace = true
clap_complete.workspace = true
clap_mangen.workspace = true
minijinja.workspace = true
schemars = { workspace = true, features = ["derive", "indexmap2"] }
serde_json = { workspace = true, features = ["preserve_order"] }
serde_yaml = { workspace = true, features = [] }
toml = { workspace = true, features = ["preserve_order"] }

[dev-dependencies]
insta = "1.43.1"
tempfile = "3.21.0"
test-case = "3.3.1"

[lints]
workspace = true

[package.metadata.generate-rpm]
assets = [
  { source = "target/release/claptrap", dest = "/usr/bin/claptrap", mode = "755" },
]
