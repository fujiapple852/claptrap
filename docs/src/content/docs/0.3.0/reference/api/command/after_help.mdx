---
title: Command::after_help
description: The `Command::after_help` API.
sidebar:
  label: after_help
slug: 0.3.0/reference/api/command/after_help
---

import { Tabs, TabItem } from '@astrojs/starlight/components';
import { Badge } from '@astrojs/starlight/components';
import { Icon } from '@astrojs/starlight/components';
import { LinkButton } from '@astrojs/starlight/components';

# `after_help` <LinkButton href="https://docs.rs/clap/latest/clap/struct.Command.html#method.after_help" icon="open-book" variant="minimal" />

<Badge text="since 0.1.0" size="small" />

Free-form help text for after auto-generated short help (`-h`).

# Example

### Spec

<Tabs syncKey="format">
  <TabItem label="TOML">
    ```toml
    name = "myprog"
    after-help = "Does really amazing things for great people... but be careful with -R!"
    [args]
    ```
  </TabItem>

  <TabItem label="YAML">
    ```yaml
    name: myprog
    after-help: Does really amazing things for great people... but be careful with -R!
    args: {}
    ```
  </TabItem>

  <TabItem label="JSON">
    ```json
    {
      "name": "myprog",
      "after-help": "Does really amazing things for great people... but be careful with -R!",
      "args": {}
    }
    ```
  </TabItem>
</Tabs>

### Input

```bash
myprog -h
```

### Output

```bash
Usage: myprog

Options:
  -h, --help  Print help

Does really amazing things for great people... but be careful with -R!
```
