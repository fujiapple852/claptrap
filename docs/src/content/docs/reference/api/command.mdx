---
title: "Command"
description: "The Command API"
---

<div class="api-table">
| API | Description |
| --- | ----------- |
| [`args`](args) | Adds multiple arguments to the list of valid possibilities. |
| [`groups`](groups) | Adds multiple `ArgGroups` to the Command at once. |
| [`subcommands`](subcommands) | Adds multiple subcommands to the list of valid possibilities. |
| [`no_binary_name`](no_binary_name) | Specifies that the parser should not assume the first argument passed is the binary name. |
| [`ignore_errors`](ignore_errors) | Try not to fail on parse errors, like missing option values. |
| [`args_override_self`](args_override_self) | Replace prior occurrences of arguments rather than error. |
| [`dont_delimit_trailing_values`](dont_delimit_trailing_values) | Disables the automatic delimiting of values after `--` or when `Arg::trailing_var_arg` was used. |
| [`color`](color) | Sets when to color output. |
| [`styles`](styles) | Sets the `Styles` for terminal output. |
| [`term_width`](term_width) | Sets the terminal width at which to wrap help messages. |
| [`max_term_width`](max_term_width) | Limit the line length for wrapping help when using the current terminal’s width. |
| [`disable_version_flag`](disable_version_flag) | Disables -V and --version flag. |
| [`propagate_version`](propagate_version) | Specifies to use the version of the current command for all subcommands. |
| [`next_line_help`](next_line_help) | Places the help string for all arguments and subcommands on the line after them. |
| [`disable_help_flag`](disable_help_flag) | Disables `-h` and `--help` flag. |
| [`disable_help_subcommand`](disable_help_subcommand) | Disables the help subcommand. |
| [`disable_colored_help`](disable_colored_help) | Disables colorized help messages. |
| [`help_expected`](help_expected) | Panic if help descriptions are omitted. |
| [`hide_possible_values`](hide_possible_values) | Tells clap not to print possible values when displaying help information. |
| [`infer_long_args`](infer_long_args) | Allow partial matches of long arguments or their aliases. |
| [`infer_subcommands`](infer_subcommands) | Allow partial matches of subcommand names and their aliases. |
| [`name`](name) | Sets the program’s name. |
| [`bin_name`](bin_name) | Overrides the runtime-determined name of the binary for help and error messages. |
| [`display_name`](display_name) | Overrides the runtime-determined display name of the program for help and error messages. |
| [`author`](author) | Sets the author(s) for the help message. |
| [`about`](about) | Sets the program’s description for the short help (`-h`). |
| [`long_about`](long_about) | Sets the program’s description for the long help (`--help`). |
| [`after_help`](after_help) | Free-form help text for after auto-generated short help (`-h`). |
| [`after_long_help`](after_long_help) | Free-form help text for after auto-generated long help (`--help`). |
| [`before_help`](before_help) | Free-form help text for before auto-generated short help (`-h`). |
| [`before_long_help`](before_long_help) | Free-form help text for before auto-generated long help (`--help`). |
| [`version`](version) | Sets the version for the short version (`-V`) and help messages. |
| [`long_version`](long_version) | Sets the version for the long version (`--version`) and help messages. |
| [`override_usage`](override_usage) | Overrides the clap generated usage string for help and error messages. |
| [`override_help`](override_help) | Overrides the clap generated help message (both `-h` and `--help`). |
| [`help_template`](help_template) | Sets the help template to be used, overriding the default format. |
| [`flatten_help`](flatten_help) | Flatten subcommand help into the current command’s help. |
| [`next_help_heading`](next_help_heading) | Set the default section heading for future args. |
| [`next_display_order`](next_display_order) | Change the starting value for assigning future display orders for args. |
| [`arg_required_else_help`](arg_required_else_help) | Exit gracefully if no arguments are present (e.g. `$ myprog`). |
| [`allow_missing_positional`](allow_missing_positional) | Allows one to implement two styles of CLIs where positionals can be used out of order. |
| [`short_flag`](short_flag) | Sets the short version of the subcommand flag without the preceding `-`. |
| [`long_flag`](long_flag) | Sets the long version of the subcommand flag without the preceding `--`. |
| [`alias`](alias) | Sets a hidden alias to this subcommand. |
| [`short_flag_alias`](short_flag_alias) | Add an alias, which functions as “hidden” short flag subcommand. |
| [`long_flag_alias`](long_flag_alias) | Add an alias, which functions as a “hidden” long flag subcommand. |
| [`aliases`](aliases) | Sets multiple hidden aliases to this subcommand. |
| [`short_flag_aliases`](short_flag_aliases) | Add aliases, which function as “hidden” short flag subcommands. |
| [`long_flag_aliases`](long_flag_aliases) | Add aliases, which function as “hidden” long flag subcommands. |
| [`visible_alias`](visible_alias) | Sets a visible alias to this subcommand. |
| [`visible_short_flag_alias`](visible_short_flag_alias) | Add an alias, which functions as “visible” short flag subcommand. |
| [`visible_long_flag_alias`](visible_long_flag_alias) | Add an alias, which functions as a “visible” long flag subcommand. |
| [`visible_aliases`](visible_aliases) | Sets multiple visible aliases to this subcommand. |
| [`visible_short_flag_aliases`](visible_short_flag_aliases) | Add aliases, which function as visible short flag subcommands. |
| [`visible_long_flag_aliases`](visible_long_flag_aliases) | Add aliases, which function as visible long flag subcommands. |
| [`display_order`](display_order) | Set the placement of this subcommand within the help. |
| [`hide`](hide) | Specifies that this subcommand should be hidden from help messages. |
| [`subcommand_required`](subcommand_required) | If no subcommand is present at runtime, error and exit gracefully. |
| [`allow_external_subcommands`](allow_external_subcommands) | Assume unexpected positional arguments are a subcommand. |
| [`external_subcommand_value_parser`](external_subcommand_value_parser) | Specifies how to parse external subcommand arguments. |
| [`args_conflicts_with_subcommands`](args_conflicts_with_subcommands) | Specifies that use of an argument prevents the use of subcommands. |
| [`subcommand_precedence_over_arg`](subcommand_precedence_over_arg) | Prevent subcommands from being consumed as an arguments value. |
| [`subcommand_negates_reqs`](subcommand_negates_reqs) | Allows subcommands to override all requirements of the parent command. |
| [`multicall`](multicall) | Multiple-personality program dispatched on the binary name (`argv[0]`). |
| [`subcommand_value_name`](subcommand_value_name) | Sets the value name used for subcommands when printing usage and help. |
| [`subcommand_help_heading`](subcommand_help_heading) | Sets the help heading used for subcommands when printing usage and help. |
</div>