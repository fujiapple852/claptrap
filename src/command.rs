use crate::arg::Arg;
use crate::arg_group::{ArgGroup, NamedArgGroup};
use crate::style::Styles;
use crate::values::ValueParser;
use indexmap::IndexMap;
use serde::{Deserialize, Deserializer};

/// Represents a command configuration for a CLI application.
#[derive(Debug, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
#[serde(rename_all = "kebab-case")]
pub struct Command {
    name: String,
    #[serde(default, deserialize_with = "deserialize_args")]
    args: Option<IndexMap<String, Arg>>,
    #[serde(default)]
    subcommands: Vec<Command>,
    groups: Option<IndexMap<String, ArgGroup>>,
    ignore_errors: Option<bool>,
    args_override_self: Option<bool>,
    dont_delimit_trailing_values: Option<bool>,
    color: Option<ColorChoice>,
    styles: Option<Styles>,
    term_width: Option<usize>,
    max_term_width: Option<usize>,
    disable_version_flag: Option<bool>,
    propagate_version: Option<bool>,
    next_line_help: Option<bool>,
    disable_help_flag: Option<bool>,
    disable_help_subcommand: Option<bool>,
    disable_colored_help: Option<bool>,
    help_expected: Option<bool>,
    hide_possible_values: Option<bool>,
    infer_long_args: Option<bool>,
    infer_subcommands: Option<bool>,
    bin_name: Option<String>,
    display_name: Option<String>,
    author: Option<String>,
    about: Option<String>,
    long_about: Option<String>,
    after_help: Option<String>,
    after_long_help: Option<String>,
    before_help: Option<String>,
    before_long_help: Option<String>,
    version: Option<String>,
    long_version: Option<String>,
    override_usage: Option<String>,
    override_help: Option<String>,
    help_template: Option<String>,
    flatten_help: Option<bool>,
    next_help_heading: Option<String>,
    next_display_order: Option<usize>,
    arg_required_else_help: Option<bool>,
    allow_missing_positional: Option<bool>,
    short_flag: Option<char>,
    long_flag: Option<String>,
    alias: Option<String>,
    short_flag_alias: Option<char>,
    long_flag_alias: Option<String>,
    aliases: Option<Vec<String>>,
    short_flag_aliases: Option<Vec<char>>,
    long_flag_aliases: Option<Vec<String>>,
    visible_alias: Option<String>,
    visible_short_flag_alias: Option<char>,
    visible_long_flag_alias: Option<String>,
    visible_aliases: Option<Vec<String>>,
    visible_short_flag_aliases: Option<Vec<char>>,
    visible_long_flag_aliases: Option<Vec<String>>,
    display_order: Option<usize>,
    hide: Option<bool>,
    subcommand_required: Option<bool>,
    allow_external_subcommands: Option<bool>,
    external_subcommand_value_parser: Option<ValueParser>,
    args_conflicts_with_subcommands: Option<bool>,
    subcommand_precedence_over_arg: Option<bool>,
    subcommand_negates_reqs: Option<bool>,
    multicall: Option<bool>,
    subcommand_value_name: Option<String>,
    subcommand_help_heading: Option<String>,
}

impl Command {
    #[must_use]
    pub fn get_name(&self) -> &str {
        &self.name
    }
    pub fn get_subcommands(&self) -> impl Iterator<Item = &Self> {
        self.subcommands.iter()
    }
}

fn deserialize_args<'de, D>(de: D) -> Result<Option<IndexMap<String, Arg>>, D::Error>
where
    D: Deserializer<'de>,
{
    let mut args = Option::<IndexMap<String, Arg>>::deserialize(de)?;
    if let Some(map) = &mut args {
        map.iter_mut().for_each(|(id, arg)| arg.id.clone_from(id));
    }
    Ok(args)
}

impl Command {
    pub fn get_arguments(&self) -> impl Iterator<Item = &Arg> {
        self.args
            .as_ref()
            .into_iter()
            .flat_map(|args| args.values())
    }
}

#[expect(clippy::cognitive_complexity, clippy::too_many_lines)]
impl From<Command> for clap::Command {
    fn from(cmd: Command) -> Self {
        let mut command = Self::new(cmd.name);
        if let Some(ignore_errors) = cmd.ignore_errors {
            command = command.ignore_errors(ignore_errors);
        }
        if let Some(args_override_self) = cmd.args_override_self {
            command = command.args_override_self(args_override_self);
        }
        if let Some(dont_delimit_trailing_values) = cmd.dont_delimit_trailing_values {
            command = command.dont_delimit_trailing_values(dont_delimit_trailing_values);
        }
        if let Some(color) = cmd.color {
            command = command.color(clap::ColorChoice::from(color));
        }
        if let Some(styles) = cmd.styles {
            command = command.styles(clap::builder::Styles::from(styles));
        }
        if let Some(term_width) = cmd.term_width {
            command = command.term_width(term_width);
        }
        if let Some(max_term_width) = cmd.max_term_width {
            command = command.max_term_width(max_term_width);
        }
        if let Some(disable_version_flag) = cmd.disable_version_flag {
            command = command.disable_version_flag(disable_version_flag);
        }
        if let Some(propagate_version) = cmd.propagate_version {
            command = command.propagate_version(propagate_version);
        }
        if let Some(next_line_help) = cmd.next_line_help {
            command = command.next_line_help(next_line_help);
        }
        if let Some(disable_help_flag) = cmd.disable_help_flag {
            command = command.disable_help_flag(disable_help_flag);
        }
        if let Some(disable_help_subcommand) = cmd.disable_help_subcommand {
            command = command.disable_help_subcommand(disable_help_subcommand);
        }
        if let Some(disable_colored_help) = cmd.disable_colored_help {
            command = command.disable_colored_help(disable_colored_help);
        }
        if let Some(help_expected) = cmd.help_expected {
            command = command.help_expected(help_expected);
        }
        if let Some(hide_possible_values) = cmd.hide_possible_values {
            command = command.hide_possible_values(hide_possible_values);
        }
        if let Some(infer_long_args) = cmd.infer_long_args {
            command = command.infer_long_args(infer_long_args);
        }
        if let Some(infer_subcommands) = cmd.infer_subcommands {
            command = command.infer_subcommands(infer_subcommands);
        }
        if let Some(bin_name) = cmd.bin_name {
            command = command.bin_name(bin_name);
        }
        if let Some(display_name) = cmd.display_name {
            command = command.display_name(display_name);
        }
        if let Some(author) = cmd.author {
            command = command.author(author);
        }
        if let Some(about) = cmd.about {
            command = command.about(about);
        }
        if let Some(long_about) = cmd.long_about {
            command = command.long_about(long_about);
        }
        if let Some(after_help) = cmd.after_help {
            command = command.after_help(after_help);
        }
        if let Some(after_long_help) = cmd.after_long_help {
            command = command.after_long_help(after_long_help);
        }
        if let Some(before_help) = cmd.before_help {
            command = command.before_help(before_help);
        }
        if let Some(before_long_help) = cmd.before_long_help {
            command = command.before_long_help(before_long_help);
        }
        if let Some(version) = cmd.version {
            command = command.version(version);
        }
        if let Some(long_version) = cmd.long_version {
            command = command.long_version(long_version);
        }
        if let Some(override_usage) = cmd.override_usage {
            command = command.override_usage(override_usage);
        }
        if let Some(override_help) = cmd.override_help {
            command = command.override_help(override_help);
        }
        if let Some(help_template) = cmd.help_template {
            command = command.help_template(help_template);
        }
        if let Some(flatten_help) = cmd.flatten_help {
            command = command.flatten_help(flatten_help);
        }
        if let Some(next_help_heading) = cmd.next_help_heading {
            command = command.next_help_heading(next_help_heading);
        }
        if let Some(next_display_order) = cmd.next_display_order {
            command = command.next_display_order(next_display_order);
        }
        if let Some(arg_required_else_help) = cmd.arg_required_else_help {
            command = command.arg_required_else_help(arg_required_else_help);
        }
        if let Some(allow_missing_positional) = cmd.allow_missing_positional {
            command = command.allow_missing_positional(allow_missing_positional);
        }
        if let Some(short_flag) = cmd.short_flag {
            command = command.short_flag(short_flag);
        }
        if let Some(long_flag) = cmd.long_flag {
            command = command.long_flag(long_flag);
        }
        if let Some(alias) = cmd.alias {
            command = command.alias(alias);
        }
        if let Some(short_flag_alias) = cmd.short_flag_alias {
            command = command.short_flag_alias(short_flag_alias);
        }
        if let Some(long_flag_alias) = cmd.long_flag_alias {
            command = command.long_flag_alias(long_flag_alias);
        }
        if let Some(aliases) = cmd.aliases {
            command = command.aliases(aliases);
        }
        if let Some(short_flag_aliases) = cmd.short_flag_aliases {
            command = command.short_flag_aliases(short_flag_aliases);
        }
        if let Some(long_flag_aliases) = cmd.long_flag_aliases {
            command = command.long_flag_aliases(long_flag_aliases);
        }
        if let Some(visible_alias) = cmd.visible_alias {
            command = command.visible_alias(visible_alias);
        }
        if let Some(visible_short_flag_alias) = cmd.visible_short_flag_alias {
            command = command.visible_short_flag_alias(visible_short_flag_alias);
        }
        if let Some(visible_long_flag_alias) = cmd.visible_long_flag_alias {
            command = command.visible_long_flag_alias(visible_long_flag_alias);
        }
        if let Some(visible_aliases) = cmd.visible_aliases {
            command = command.visible_aliases(visible_aliases);
        }
        if let Some(visible_short_flag_aliases) = cmd.visible_short_flag_aliases {
            command = command.visible_short_flag_aliases(visible_short_flag_aliases);
        }
        if let Some(visible_long_flag_aliases) = cmd.visible_long_flag_aliases {
            command = command.visible_long_flag_aliases(visible_long_flag_aliases);
        }
        if let Some(display_order) = cmd.display_order {
            command = command.display_order(display_order);
        }
        if let Some(hide) = cmd.hide {
            command = command.hide(hide);
        }
        if let Some(subcommand_required) = cmd.subcommand_required {
            command = command.subcommand_required(subcommand_required);
        }
        if let Some(allow_external_subcommands) = cmd.allow_external_subcommands {
            command = command.allow_external_subcommands(allow_external_subcommands);
        }
        if let Some(external_subcommand_value_parser) = cmd.external_subcommand_value_parser {
            command = command.external_subcommand_value_parser(clap::builder::ValueParser::from(
                external_subcommand_value_parser,
            ));
        }
        if let Some(args_conflicts_with_subcommands) = cmd.args_conflicts_with_subcommands {
            command = command.args_conflicts_with_subcommands(args_conflicts_with_subcommands);
        }
        if let Some(subcommand_precedence_over_arg) = cmd.subcommand_precedence_over_arg {
            command = command.subcommand_precedence_over_arg(subcommand_precedence_over_arg);
        }
        if let Some(subcommand_negates_reqs) = cmd.subcommand_negates_reqs {
            command = command.subcommand_negates_reqs(subcommand_negates_reqs);
        }
        if let Some(multicall) = cmd.multicall {
            command = command.multicall(multicall);
        }
        if let Some(subcommand_value_name) = cmd.subcommand_value_name {
            command = command.subcommand_value_name(subcommand_value_name);
        }
        if let Some(subcommand_help_heading) = cmd.subcommand_help_heading {
            command = command.subcommand_help_heading(subcommand_help_heading);
        }
        command = command.args(if let Some(args) = cmd.args {
            args.into_values().map(clap::Arg::from).collect::<Vec<_>>()
        } else {
            Vec::new()
        });
        if let Some(groups) = cmd.groups {
            command = command.groups(
                groups
                    .into_iter()
                    .map(|(name, group)| clap::ArgGroup::from(NamedArgGroup::new(name, group))),
            );
        }
        command = command.subcommands(cmd.subcommands);
        command
    }
}

#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum ColorChoice {
    Auto,
    Always,
    Never,
}

impl From<ColorChoice> for clap::ColorChoice {
    fn from(color_choice: ColorChoice) -> Self {
        match color_choice {
            ColorChoice::Auto => Self::Auto,
            ColorChoice::Always => Self::Always,
            ColorChoice::Never => Self::Never,
        }
    }
}
