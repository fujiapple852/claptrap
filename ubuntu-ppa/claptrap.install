target/release/claptrap usr/bin
