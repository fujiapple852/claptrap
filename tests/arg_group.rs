use claptrap::command::Command;
use claptrap::parse;
use std::ffi::OsString;

#[test]
fn test_args() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"] }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
        "#,
    )
    .unwrap();
    let input = "-f";
    let args: Vec<OsString> = input.split(' ').map(OsString::from).collect();
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}

#[test]
fn test_multiple() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"], multiple = true }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
        "#,
    )
    .unwrap();
    let input = "-f -c";
    let args: Vec<OsString> = input.split(' ').map(OsString::from).collect();
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}

#[test]
fn test_required() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"], required = true }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
        "#,
    )
    .unwrap();
    let args: Vec<OsString> = vec![];
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}

#[test]
fn test_requires() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"], requires = "debug" }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
            debug = { short = 'd', action = "set-true" }
        "#,
    )
    .unwrap();
    let input = "-c";
    let args: Vec<OsString> = input.split(' ').map(OsString::from).collect();
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}

#[test]
fn test_requires_all() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"], requires-all = ["debug", "verb"] }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
            debug = { short = 'd', action = "set-true" }
            verb = { short = 'v', action = "set-true" }
        "#,
    )
    .unwrap();
    let input = "-c -d";
    let args: Vec<OsString> = input.split(' ').map(OsString::from).collect();
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}

#[test]
fn test_conflicts_with() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"], conflicts-with = "debug" }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
            debug = { short = 'd', action = "set-true" }
        "#,
    )
    .unwrap();
    let input = "-c -d";
    let args: Vec<OsString> = input.split(' ').map(OsString::from).collect();
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}

#[test]
fn test_conflicts_with_all() {
    let app: Command = toml::from_str(
        r#"
            name = "myprog"
            [groups]
            req_flags = { args = ["flag", "color"], conflicts-with-all = ["debug", "verb"] }
            [args]
            flag = { short = 'f', action = "set-true" }
            color = { short = 'c', action = "set-true" }
            debug = { short = 'd', action = "set-true" }
            verb = { short = 'v', action = "set-true" }
        "#,
    )
    .unwrap();
    let input = "-c -v";
    let args: Vec<OsString> = input.split(' ').map(OsString::from).collect();
    let output = parse(app, args);
    insta::assert_snapshot!(output);
}
