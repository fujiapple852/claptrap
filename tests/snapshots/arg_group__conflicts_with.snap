---
source: tests/arg_group.rs
expression: "util::run(spec, args)"
---
command cat <<'EOF_224b398c'
[1m[31merror:[0m the argument '[33m-c[0m' cannot be used with '[33m-d[0m'

[1m[4mUsage:[0m [1mmyprog[0m [1m-c[0m

For more information, try '[1m--help[0m'.
EOF_224b398c
exit 1
